### ISSUE TITLE
Analyse x.__bool__() <=> bool(x) for reachabilty  

### ISSUE THREAD

**Issue by Gobot1234** [microsoft/pyright#5515]
**Is your feature request related to a problem? Please describe.**
Currently objects which are marked as returning Literal[False/True] do not actually have their return types used for reachabilty analysis, there is also no way to get the return type using `bool(x)` which should act an alias for `x.__bool__()` (or maybe `type(x).__bool__(x)` if you'd prefer 😉)
 
**Describe the solution you'd like**
I'd like to have the return of `__bool__` be used for type checking and also have `bool()` call the methods bool to get the return type and not just have it be `bool` when it could be more specific.

**Additional context**
MyPy is implementing a similar feature and I've found it very successful at finding bugs https://github.com/python/mypy/pull/15645

**Comment by erictraut** [comment]
This seems like something that should be handled in the typeshed stubs, not in special-case logic within a type checker. The `bool.__new__` method is currently defined as:

```python
    def __new__(cls, __o: object = ...) -> Self: ...
```

It could be defined as an overload that returns `Literal[True]` or `Literal[False]` if the supplied object implements a `__bool__` method that returns a literal value.

**Comment by Gobot1234** [comment]
I think pyright has some issues with defining the __new__ method of bool like that

```py
# builtins.pyi
class _Truthy(Protocol):
    def __bool__(self) -> Literal[True]: ...

class _Falsy(Protocol):
    def __bool__(self) -> Literal[False]: ...

@final
class bool(int):
    @overload
    def __new__(cls, __o: _Truthy) -> Literal[True]: ...
    @overload
    def __new__(cls, __o: _Falsy) -> Literal[False]: ...
    @overload
    def __new__(cls, __o: object = ...) -> bool: ...
```

```py
# test.py
class MissingSentinel(Any if TYPE_CHECKING else object):
    __slots__ = ()

    def __eq__(self, other: object) -> Literal[False]:
        return False

    def __bool__(self) -> Literal[False]:
        return False

    def __hash__(self) -> Literal[0]:
        return 0

    def __repr__(self) -> Literal["..."]:
        return "..."

MISSING: Final = MissingSentinel()
reveal_type(MISSING.__bool__)  # type is () -> Literal[False]
reveal_type(bool(MISSING.__bool__()))  # revealed type is bool?
reveal_type(bool(MISSING))  # revealed type is Literal[True]???
```

**Comment by Gobot1234** [comment]
This appears to be caused by the weird bases but I'll run this change by the typeshed folk cause it seems to mostly work apart from this

**Comment by Gobot1234** [comment]
Also adding to this the types of `True/False.__bool__()` isn't a Literal which I think could be improved

**Comment by erictraut** [comment]
When you define a class that derives from `Any`, many things won't work correctly. Out of curiosity, why are you doing that? If you delete `Any if TYPE_CHECKING else object`, you'll get better results.

**Comment by Gobot1234** [comment]
This class is for a private sentinel that shouldn't be exposed to users and can be assigned to any type where things like None don't make sense as a default.
